Cyptography pr 20566 rebase

abadger opened the pull request:
##### SUMMARY
Port the ansible code to work with the cryptography python library as this is the library that everyone in the python community is getting behind for cryptographic functions.

We maintain a fallback for pycrypto so that older distros will still work out of the box.

The idea is for the following to work:
* ```pip install ansible``` to install with cryptography as its setuptools dep (install time and runtime)
* ```ANSIBLE_CRYPTO_BACKEND=pycrypto pip install ansible``` will install with pycrypto as its setuptools  dep (installtime and runtime).
* If both cryptography and pycrypto are installed, ansible will prefer to use cryptography whether or not it is the setuptools dep.


##### ISSUE TYPE
 - Feature Pull Request
 - Bugfix Pull Request

##### COMPONENT NAME
primarily vault

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
devel (2.4)
```


##### ADDITIONAL INFORMATION
based on https://github.com/ansible/ansible/pull/24894

Which in turn is based on https://github.com/ansible/ansible/pull/20566

@reaperhulk, @alikins, Please have a look for the crypto code.  I'll see if we can wrangle some security people to look at it as well.

@mattclay, @gundalow Please have a look for the test portion.  docker images probably need to be updated.  It would be good to find a way to test the code when using both cryptography and when using pycrypto.  We might be able to do that by monkeypatching HAS_CRYPTOGRAPHY to False when we want to choose the pycrypto code (This is assuming that the test machine has both installed).

reaperhulk committed:
Replace PyCrypto with pyca/cryptography

pyca/cryptography is already implicitly a dependency in many cases
through paramiko (2.0+) as well as the new openssl_publickey module,
which requires pyOpenSSL 16.0+. Additionally, pyca/cryptography is
an optional dep for better performance with vault already.

This commit leverages cryptography's padding, constant time comparisons,
and CBC/CTR modes to reduce the amount of code ansible needs to
maintain.

reaperhulk committed:
change various requirements.txts I missed previously

reaperhulk committed:
update dependencies for various CI scripts

reaperhulk committed:
additional CI dockerfile/script updates

reaperhulk committed:
add paramiko to the windows requirement set

This is needed because ansible lists it as a requirement. Previously
the missing dep wasn't enforced, but cryptography imports pkg_resources
so you can't ignore a requirement any more

reaperhulk committed:
add paramiko to sanity list

This requirement is again because ansible lists paramiko as a hard dep.
This was not previously enforced but cryptography imports pkg_resources
from setuptools and now it is.

nicoonoclaste committed:
contrib/inventory/gce: Remove spurious require on pycrypto

(cherry picked from commit 9e16b9db275263b3ea8d1b124966fdebfc9ab271)

alikins committed:
merge resolution cleanup

alikins committed:
merge fixup on vault/__init__.py (ciphertext)

alikins committed:
remove now unused check_prereqs call in encrypt_bytes

alikins committed:
Add intg test cases for old vault and for wrong passwords

abadger committed:
Change hacking README for dual pycrypto/cryptography

abadger committed:
Add back in atfork since we need pycrypto to reinitialize its RNG jsut in case we're being used with old paramiko

abadger committed:
Add cryptography to ec2_win_password requirements

abadger committed:
Allow vault to use either cryptography and pycrypto

The code will prefer cryptography if installed but fall back to pycrypto
if not

alikins committed:
helper script for manual testing of pycrypto/cryptography

abadger committed:
Readd tests for pycrypto related functions

The main unittests are still testing cryptography exclusively (assuming
that you have both installed) but readdd the tests for pycrypto helper
functions.  Also go back to installing pycrypto for test docker images.
This is in preparation for testing code under both pycrypto and
cryptography

abadger committed:
Attempt to add version or setup

abadger committed:
No longer need to install cryptography separately

abadger committed:
Fix call to _decrypt_cryptography in VaultAES

abadger committed:
Update test call to use implementation specific create_key calls

abadger committed:
Another problem with the calling of the new create_key functions

abadger committed:
Do not display deprecation warning on python-2.6

abadger committed:
More fixes to the test's calling of create_key functions

abadger committed:
Namespace all of the pycrypto imports and always import them

Makes unittests better and the code less likely to get stupid mistakes
(like using HMAC from cryptogrpahy when the one from pycrypto is needed)

abadger committed:
Handle wrong password given for VaultAES format

abadger committed:
Handle pycrypto imports with namespace correctly

abadger commented:
Okay, integration tests passing locally (on Fedora 25) with:
```
 test/runner/ansible-test integration  --python 3.5 vault
 test/runner/ansible-test integration  --python 2.7 vault
```

unittests passing locally with
```
test/runner/ansible-test units --python 3.5 
test/runner/ansible-test units --python 2.7
```

The unittests require that both pycrypto and cryptography are installed which is not the case in the docker images.  what do we want to do about that? Options:

* Install both cryptography and pycrypto
* Rewrite the unittests to require only cryptography and update the docker images (easy to rewrite as it's just removing some calls that specifically target pycrypto)
* Have certain unittests skip if pycrypto isn't installed.

abadger committed:
One more fix to the unittests

abadger committed:
Skip the pycrypto tests so that users without it installed can still run the unittests

abadger committed:
Run unittests for vault with both cryptography and pycrypto backend

abadger commented:
Okay, unittests now are duplicated for cryptography and pycrypto as their backend and only run condtionally depending on which is installed.  @reaperhulk, I also found out why you had to add paramiko when we didn't.  cryptography is using pkg_resources at runtime.  Because of that, it forces our deps to be available too.  Since paramiko wasn't available, we go boom.  pkg_resources is unfortunately fragile like this (we have to use it too but for a different reason).  Since you're only using it for loading plugins, you might want to look into something that's more robust like straight.plugins.

reaperhulk commented on lib/ansible/modules/cloud/amazon/ec2_win_password.py:
```diff
@@ -153,15 +161,12 @@ def main():
             with f:
-                key = RSA.importKey(f.read(), key_passphrase)
-        except (ValueError, IndexError, TypeError) as e:
+                key = load_pem_private_key(f.read(), key_passphrase, BACKEND)
```
This will fail on Python 3 as f is opened as 'r' rather than 'rb' and load_pem_private_key requires bytes. key_passphrase must also be bytes -- if that path accepts utf8 `encode('utf8')` would be appropriate

abadger replied:
Good point.  I'll convert this to use 'rb'.

reaperhulk commented on test/integration/targets/vault/vault-password-wrong:
```diff
@@ -0,0 +1 @@
+hunter42
```
😆

abadger committed:
PEP8 fixes

reaperhulk commented on lib/ansible/parsing/vault/__init__.py:
```diff
@@ -704,104 +723,139 @@ class VaultAES256:
+    def _encrypt_cryptography(b_plaintext, b_salt, b_key1, b_key2, b_iv):
+        cipher = C_Cipher(algorithms.AES(b_key1), modes.CTR(b_iv), CRYPTOGRAPHY_BACKEND)
+        encryptor = cipher.encryptor()
+        padder = padding.PKCS7(algorithms.AES.block_size).padder()
```
CTR mode doesn't actually require padding, so this is unnecessary. That said, I assume you're staying compatible with existing vault implementations which already do this. It's not a security thing, just a few wasted bytes/CPU cycles.

bcoca replied:
yes, we've been wanting to change that for a while but were waiting until we made the architecture more pluggable to allow for old/new formats to be used transparently.

reaperhulk reviewed (commented):
Thanks for taking this over and doing so much work! I've put a few comments on it after review.

One thing that might be useful to know: cryptography is planning on beginning to ship a manylinux1 wheel soon(-ish). When that happens ansible users on manylinux1 compatible distributions will no longer need to install OpenSSL development headers. cffi already ships a manylinux1 wheel, which, if the user is on pip 8.1+, obviates the need for libffi dev headers.

jimi-c commented on lib/ansible/parsing/vault/__init__.py:
```diff
@@ -704,104 +723,139 @@ class VaultAES256:
+            b_iv = hexlify(b_derivedkey[(key_length * 2):(key_length * 2) + iv_length])
         else:
-            b_derivedkey = cls._create_key(b_password, b_salt, keylength, ivlength)
+            display.warning(u"Neither Cryptography nor pycrypto was installed. This should have been detected earlier")
```
Is this really required?

abadger replied:
It shouldn't be.... But I felt like I needed to put something in here because we (alikins and I) think that all of these should be caught in __init__().  Not sure how we can signal this otherwise....

abadger committed:
ec2_win_password changes

* Add note about cryptography now being required
* Fix python3 bug which would pass text strings to a function which
  requires byte strings.

abadger committed:
* Target deprecationwarning specifically and for much fewer imports when dealing with cryptography in vault

abadger committed:
Split pycrypto from version constraint

abadger committed:
Merge warning into AnsibleError message

abadger committed:
Fix name of new internal function

abadger committed:
Another fix for refactored hash function

abadger committed:
Don't see any reason for a closure here so making it it's own private method.

abadger commented:
Thanks @jimi-c @bcoca for the reviews.  @reaperhulk, thanks for the review and the bulk of the code!  The information about cryptography gaining a manylinux wheel soon should also be good news for us going forward.

I think that I've addressed all the issues you've collectively brought up. The one outstanding question I see is whether the integration tests will work with just the changes here once the docker images are updated.  I'm hopeful since integration tests are passing here but we shall have to see.  I'm also hoping that mckerrj and I can find someone on the redhat security team to review this code before we put it in but even if that doesn't happen, a post-merge review should still be doable.

I'm going on vacation followed by ansiblefest in London and will be gone for the next two weeks.  (Can check in on this sporadically, though). I have no objections if anyone else wants to work on this for any furthe rissues that come up in that time.

nitzmahone commented on hacking/README.md:
```diff
@@ -17,7 +17,7 @@ and do not wish to install them from your operating system package manager, you
 can install them from pip
 
     $ easy_install pip               # if pip is not already available
```
Unrelated, but wonder if we should ditch reference to easy_install, as pip now recommends `get-pip.py`...

gundalow replied:
This may need changing in other places, such as install guide.

abadger replied:
I think we should do this one separately.  There's questions of  availability that we'd need to figure out. (I checked whether I had get_pip.py on a centos6 box and it wasn't there... then I checked fedora25 and it wasn't installed either).

abadger replied:
Okay, on python-2.7.13 and python-3.5 (have to check older versions for pyhton2 but I'm pretty sure it wasn't in python-2.6.x) you can run this:
```
$ python -m ensurepip
```

to get pip and dependencies installed.

reaperhulk replied:
ensurepip was added in 3.4+ and backported to 2.7 in 2.7.9+

nitzmahone commented on lib/ansible/parsing/vault/__init__.py:
```diff
@@ -20,47 +20,56 @@
-    HAS_COUNTER = True
-except ImportError:
-    HAS_COUNTER = False
+    from Crypto.Cipher import AES as pycrypto_AES
```
Minor consistency nit- our stuff seems to be using `[action]_[backend]()`, where these aliased imports have that reversed. I'd vote for `[action]_[backend]()` over this- makes the code a little easier to read...

nitzmahone commented on lib/ansible/parsing/vault/__init__.py:
```diff
@@ -73,37 +82,17 @@
-        raise AnsibleError(CRYPTO_UPGRADE)
+NEED_CRYPTO_LIBRARY = "ansible-vault requires either the cryptography library (preferred) or"
+if HAS_SOME_PYCRYPTO:
+    NEED_CRYPTO_LIBRARY += " a newer version of"
```
Do we know a minimum version we could specify?

abadger replied:
It may be pycrypto 2.6 but I'm not sure.  2.6 is the minimum version that was specified in the requirements.txt file (in code in setup.py after this patch).  If you want to do some research and update the patch.... The message as written isn't a regression, though...

gundalow commented:
Docker images have been updated, closing and re-opening to rerun CI

nitzmahone reviewed (commented):
LGTM, couple of minor nits that may or may not be worth addressing...

abadger committed:
Reversing the order of namespace elements in pycrypto imports.

abadger commented:
Failures and what's probably wrong:
* freebsd-10; looks like the instance went away, not the PRs fault.
* ubuntu 14, 16, 16/py3 all need to have gcc added to the docker images.

mattclay commented:
The docker images have been updated and CI is now passing.

abadger commented:
/me will merge tomorrow when I'll be around all day in case some unexpected error arises from this.

abadger commented:
Merged to devel for 2.4.0 inclusion

abadger commented:
Thanks to @reaperhulk @alikins @nbraud and the many reviewers and testers for bringing this PR to completion!
